File: Day 14/solution.py
"""
Day 14 solution.
"""
import enum
import sys
from typing import Sequence


class Solver:
    """Day 14 solver."""

    def __init__(self, filepath: str = "input.txt"):
        self.filepath = filepath

    class symbols(enum.Enum):
        """Valid input symbols."""

        SPACES = "."
        ROUND_ROCKS = "O"
        CUBED_ROCKS = "#"

    def parse_file(
        self,
    ) -> tuple[list[list[int]], list[list[int]], int]:
        """
        Parse the file to get the round rocks and the cubed rocks.
        """
        round_rocks = []
        cubed_rocks = []
        n = 0
        with open(
            self.filepath, "r", encoding=sys.getdefaultencoding()
        ) as file:
            for i, line in enumerate(file):
                round_rocks.append([])
                cubed_rocks.append([])

                for j, c in enumerate(line.strip()):
                    if c == self.symbols.SPACES.value:
                        continue

                    if c == self.symbols.ROUND_ROCKS.value:
                        round_rocks[i].append(j)
                    elif c == self.symbols.CUBED_ROCKS.value:
                        cubed_rocks[i].append(j)
                    else:
                        raise ValueError(
                            f"Expected chars to either be '.', '#', or 'O', got '{c}'."
                        )
                n = len(line.strip())
        return round_rocks, cubed_rocks, n

    @staticmethod
    def transpose_row_list(
        row_list: list[list[int]], new_length: int
    ) -> list[list[int]]:
        """Transpose the graph."""
        if not row_list:
            raise ValueError("Cannot transpose empty row list.")

        result = [[] for _ in range(new_length)]
        for i, row in enumerate(row_list):
            for j in row:
                result[j].append(i)
        return result

    @staticmethod
    def tilt(
        round_rocks: list[list[int]],
        cubed_rocks: list[list[int]],
        num_cols: int,
        transpose: bool,
        in_reverse: bool,
    ) -> list[list[int]]:
        """Tilts the map."""
        num_rows = len(round_rocks)
        new_round_rocks = [[] for _ in range(num_rows)]

        if transpose:
            round_rocks = Solver.transpose_row_list(round_rocks, num_cols)
            cubed_rocks = Solver.transpose_row_list(cubed_rocks, num_cols)

        for axis_value, (rounds, cubes) in enumerate(
            zip(round_rocks, cubed_rocks)
        ):
            i = 0
            last = -1
            if in_reverse:
                i = len(cubes) - 1
                last = num_rows if transpose else num_cols
                rounds = reversed(rounds)

            for round in rounds:
                if in_reverse:
                    while i >= 0 and cubes[i] > round:
                        last = cubes[i]
                        i -= 1
                    last -= 1
                else:
                    while i < len(cubes) and cubes[i] < round:
                        last = cubes[i]
                        i += 1
                    last += 1

                new_round_rocks[last if transpose else axis_value].append(
                    axis_value if transpose else last
                )

        return new_round_rocks

    @staticmethod
    def tilt_north(
        round_rocks: list[list[int]],
        cubed_rocks: list[list[int]],
        num_cols: int,
    ) -> list[list[int]]:
        """Tilt the map north."""
        return Solver.tilt(round_rocks, cubed_rocks, num_cols, True, False)

    @staticmethod
    def tilt_west(
        round_rocks: list[list[int]],
        cubed_rocks: list[list[int]],
        num_cols: int,
    ) -> list[list[int]]:
        """Tilt the map west."""
        return Solver.tilt(round_rocks, cubed_rocks, num_cols, False, False)

    @staticmethod
    def tilt_south(
        round_rocks: list[list[int]],
        cubed_rocks: list[list[int]],
        num_cols: int,
    ) -> list[list[int]]:
        """Tilt the map south."""
        return Solver.tilt(round_rocks, cubed_rocks, num_cols, True, True)

    @staticmethod
    def tilt_east(
        round_rocks: list[list[int]],
        cubed_rocks: list[list[int]],
        num_cols: int,
    ) -> list[list[int]]:
        """Tilt the map east."""
        return Solver.tilt(round_rocks, cubed_rocks, num_cols, False, True)

    @staticmethod
    def row_list_to_key(graph: list[list[int]]) -> tuple[tuple[int, ...], ...]:
        """Convert the graph to a hashable key."""
        return tuple(tuple(row) for row in graph)

    @staticmethod
    def calculate_north_load(round_rocks: Sequence[Sequence[int]]) -> int:
        """Calculate the total north load of the current round rock configuration."""
        return sum(
            len(row) * (len(round_rocks) - i)
            for i, row in enumerate(round_rocks)
        )

    def part_1(self) -> int:
        """Part 1 solver."""
        round_rocks, cubed_rocks, num_cols = self.parse_file()

        round_rocks = self.tilt_north(round_rocks, cubed_rocks, num_cols)
        return self.calculate_north_load(round_rocks)

    def part_2(self) -> int:
        """Part 2 solver."""
        round_rocks, cubed_rocks, num_cols = self.parse_file()

        seen: dict[tuple[tuple[int, ...], ...], int] = {}
        round_to_key: list[tuple[tuple[int, ...], ...]] = []
        cycle = cycle_start = -1

        for k in range(1000000000):
            key = self.row_list_to_key(round_rocks)
            if key in seen:
                cycle_start = seen[key]
                cycle = k - seen[key]
                break
            seen[key] = k
            round_to_key.append(key)

            for tilt in (
                self.tilt_north,
                self.tilt_west,
                self.tilt_south,
                self.tilt_east,
            ):
                round_rocks = tilt(round_rocks, cubed_rocks, num_cols)

        if cycle != -1 and cycle_start != -1:
            round_rocks = round_to_key[
                cycle_start + (1000000000 - cycle_start) % cycle
            ]

        return self.calculate_north_load(round_rocks)

    def solve(self) -> None:
        """Runs part 1 and part 2."""
        print("Day 14")
        print(f"Solving: {self.filepath}")
        print("Part 1:")
        print(self.part_1())
        print("---")
        print("Part 2:")
        print(self.part_2())


if __name__ == "__main__":
    match sys.argv:
        case [_]:
            solver = Solver()
        case [_, "--path", path] | [_, "-p", path] | [_, path]:
            solver = Solver(path)
        case _:
            print("Usage: python solution.py <path>")
            sys.exit(0)
    solver.solve()

File: Day 14/solution.cpp
#include <filesystem>
#include <fstream>
#include <iostream>
#include <vector>
#include <functional>
#include <utility>

enum symbols
{
	SPACES = '.',
	ROUND_ROCKS = 'O',
	CUBED_ROCKS = '#'
};

typedef std::vector<std::vector<int>> row_list;

class Solver
{
	std::filesystem::path filepath;

	std::tuple<row_list, row_list, int> parse_file()
	{
		row_list round_rocks, cubed_rocks;
		int num_cols = 0, i = 0;
		std::ifstream file(this->filepath);
		while (!file.eof())
		{
			round_rocks.emplace_back();
			cubed_rocks.emplace_back();
			int j = 0;
			for (char c = file.get(); c != '\n' && c != EOF; c = file.get(), ++j)
			{
				if (c == SPACES)
					continue;

				if (c == ROUND_ROCKS)
				{
					round_rocks[i].push_back(j);
				}
				else if (c == CUBED_ROCKS)
				{
					cubed_rocks[i].push_back(j);
				}
				else
				{
					throw "Expected chars to be either '.', '#', or '0'.";
				}
			}
			num_cols = j;
			++i;
		}
		file.close();
		return std::make_tuple(round_rocks, cubed_rocks, num_cols);
	}

	row_list transpose_row_list(const row_list &rows, int new_length)
	{
		if (rows.empty())
			throw "Cannot transpose empty row list.";

		row_list result(new_length);
		for (int i = 0; i < rows.size(); ++i)
		{
			for (int j : rows[i])
				result[j].push_back(i);
		}

		return result;
	}

	row_list tilt(row_list &round_rocks, row_list &cubed_rocks, int num_cols, bool transpose, bool in_reverse)
	{
		int num_rows = round_rocks.size();
		row_list new_round_rocks(num_rows);
		std::function<void(int)> traverse;
		if (in_reverse)
		{
			traverse = [&round_rocks = std::as_const(round_rocks),
									&cubed_rocks = std::as_const(cubed_rocks),
									&new_round_rocks,
									transpose,
									num_rows,
									num_cols](int axis_value)
			{
				int i = cubed_rocks[axis_value].size() - 1,
						last = transpose ? num_rows : num_cols;
				std::vector<int> rounds = round_rocks[axis_value],
												 cubes = cubed_rocks[axis_value];

				for (auto it = rounds.rbegin(); it != rounds.rend(); ++it)
				{
					int round = *it;
					while (i >= 0 && cubes[i] > round)
					{
						last = cubes[i--];
					}
					--last;
					new_round_rocks[transpose ? last : axis_value].push_back(
							transpose ? axis_value : last);
				}
			};
		}
		else
		{
			traverse = [&round_rocks = std::as_const(round_rocks),
									&cubed_rocks = std::as_const(cubed_rocks),
									&new_round_rocks,
									transpose](int axis_value)
			{
				int i = 0, last = -1;
				std::vector<int> rounds = round_rocks[axis_value],
												 cubes = cubed_rocks[axis_value];
				for (int round : rounds)
				{
					while (i < cubes.size() && cubes[i] < round)
					{
						last = cubes[i++];
					}
					++last;

					new_round_rocks[transpose ? last : axis_value].push_back(
							transpose ? axis_value : last);
				}
			};
		}

		if (transpose)
		{
			round_rocks = this->transpose_row_list(round_rocks, num_cols);
			cubed_rocks = this->transpose_row_list(cubed_rocks, num_cols);
		}

		for (int axis_value = 0; axis_value < round_rocks.size(); ++axis_value)
		{
			traverse(axis_value);
		}

		if (transpose)
		{
			// Fixes the mutation from the initial transpose.
			round_rocks = this->transpose_row_list(round_rocks, num_cols);
			cubed_rocks = this->transpose_row_list(cubed_rocks, num_cols);
		}

		return new_round_rocks;
	}

	row_list tilt_north(row_list &round_rocks, row_list &cubed_rocks, int num_cols)
	{
		return this->tilt(round_rocks, cubed_rocks, num_cols, true, false);
	}

	row_list tilt_west(row_list &round_rocks, row_list &cubed_rocks, int num_cols)
	{
		return this->tilt(round_rocks, cubed_rocks, num_cols, false, false);
	}

	row_list tilt_south(row_list &round_rocks, row_list &cubed_rocks, int num_cols)
	{
		return this->tilt(round_rocks, cubed_rocks, num_cols, true, true);
	}

	row_list tilt_east(row_list &round_rocks, row_list &cubed_rocks, int num_cols)
	{
		return this->tilt(round_rocks, cubed_rocks, num_cols, false, true);
	}

	int calculate_north_load(const row_list &round_rocks)
	{
		int total = 0;
		for (int i = 0; i < round_rocks.size(); ++i)
		{
			total += round_rocks[i].size() * (round_rocks.size() - i);
		}
		return total;
	}

	std::string row_list_to_key(const row_list &graph)
	{
		std::string result = "";

		for (const std::vector<int> &row : graph)
		{
			result += "(";
			for (int x : row)
				result += std::to_string(x) + ",";

			result += ")";
		}
		return result;
	}

	int part_1()
	{
		auto [round_rocks, cubed_rocks, num_cols] = this->parse_file();
		round_rocks = this->tilt_north(round_rocks, cubed_rocks, num_cols);

		return this->calculate_north_load(round_rocks);
	}

	int part_2()
	{
		auto [round_rocks, cubed_rocks, num_cols] = this->parse_file();

		std::unordered_map<std::string, int> seen;
		std::vector<std::vector<std::vector<int>>> round_to_key;
		int cycle = -1, cycle_start = -1, target_cycles = 1000000000;

		for (int k = 0; k < target_cycles; ++k)
		{
			std::string key = this->row_list_to_key(round_rocks);
			if (seen.contains(key))
			{
				cycle_start = seen[key];
				cycle = k - cycle_start;
				break;
			}
			seen[key] = k;
			round_to_key.push_back(round_rocks);

			round_rocks = this->tilt_north(round_rocks, cubed_rocks, num_cols);
			round_rocks = this->tilt_west(round_rocks, cubed_rocks, num_cols);
			round_rocks = this->tilt_south(round_rocks, cubed_rocks, num_cols);
			round_rocks = this->tilt_east(round_rocks, cubed_rocks, num_cols);
		}

		if (cycle != -1 && cycle_start != -1)
		{
			round_rocks = round_to_key[cycle_start + (target_cycles - cycle_start) % cycle];
		}

		return this->calculate_north_load(round_rocks);
	}

public:
	Solver(std::filesystem::path filepath = "input.txt") : filepath(filepath) {}

	void solve()
	{
		std::cout << "Day 14" << std::endl;
		std::cout << "Solving: " << this->filepath << std::endl;
		std::cout << "Part 1:" << std::endl;
		std::cout << this->part_1() << std::endl;
		std::cout << "---" << std::endl;
		std::cout << "Part 2:" << std::endl;
		std::cout << this->part_2() << std::endl;
	}
};

int main(int argc, char *argv[])
{
	std::unique_ptr<Solver> solver;
	if (argc == 1)
	{
		solver = std::make_unique<Solver>();
	}
	else if (argc == 2)
	{
		solver = std::make_unique<Solver>(argv[1]);
	}
	else if (argc == 3 && (argv[1] == "-p" || argv[1] == "--path"))
	{
		solver = std::make_unique<Solver>(argv[2]);
	}
	else
	{
		return 1;
	}

	solver->solve();
}